First draft of http package

markstory opened the pull request:
I've had to move an exception around and add a new interface to allow separation between the controller and http interfaces. While there is one reference to `Router` in `ActionDispatcher` the `class_exists()` check should make it an optional dependency.

markstory committed:
First draft of http package

I've had to move an exception around and add a new interface to allow
separation between the controller and http interfaces. While there is
one reference to `Router` in `ActionDispatcher` the `class_exists()`
check should make it an optional dependency.

saeideng commented on src/Http/composer.json:
```diff
@@ -0,0 +1,40 @@
+        "issues": "https://github.com/cakephp/cakephp/issues",
+        "forum": "https://stackoverflow.com/tags/cakephp",
+        "irc": "irc://irc.freenode.org/cakephp",
+        "source": "https://github.com/cakephp/form"
```
```suggestion
        "source": "https://github.com/cakephp/http"
```

saeideng commented on src/Http/composer.json:
```diff
@@ -0,0 +1,40 @@
+    "authors": [
+        {
+            "name": "CakePHP Community",
+            "homepage": "https://github.com/cakephp/form/graphs/contributors"
```
```suggestion
            "homepage": "https://github.com/cakephp/http/graphs/contributors"
```

ADmad commented on src/Http/ControllerInterface.php:
```diff
@@ -0,0 +1,73 @@
+     *
+     * @return bool
+     */
+    public function isAutoRenderEnabled(): bool;
```
This method feels quite out of place on the interface.

markstory replied:
I agree but the ActionDispatcher does call this method so it was necessary to get typing to work.

ADmad replied:
Couldn't the called checking using `isAutoRenderEnabled()` and the rendering be moved inside `Controller::invokeAction()` itself?

markstory replied:
Yes, that could work. We could move the isAutoRenderEnabled() and render calls there.

markstory replied:
Done :+1:

markstory committed:
Apply suggestions from code review

Co-Authored-By: saeideng <saeideng@users.noreply.github.com>

markstory committed:
Fix references to MissingControllerException

garas commented on src/Http/Exception/MissingControllerException.php:
```diff
@@ -0,0 +1,43 @@
+     * @param int|null $code The code of the error, is also the HTTP status code for the error. Defaults to 404.
+     * @param \Throwable|null $previous the previous exception.
+     */
+    public function __construct($message, ?int $code = 404, ?Throwable $previous = null)
```
I've looked why there is `404` here, but `500` in `ExceptionsTest::exceptionProvider()`.

It should be no `__construct()` here, but it should be `protected $_defaultCode = 404;` instead. With second argument as `null`, it fallbacks to `500`.

Core exceptions has mixed between `__construct()` and `protected $_defaultCode`, they should be probably normalized in next PR.

markstory committed:
Fix phpcs and psalm errors.

markstory committed:
Fix default error code for MissingControllerException

markstory committed:
Ignore this redundant condition here

The routing package is an optional dependency and psalm can't know that.

markstory committed:
Narrow ControllerInterface

Move the isAutoRenderEnabled() logic into Controller::invokeAction()
this allows the Http package to have a much smaller and more general
interface to controllers.

markstory commented:
@ADmad Anything else you'd like to see here?

markstory commented on src/Http/ControllerInterface.php:
```diff
@@ -0,0 +1,66 @@
+     * @throws \Cake\Controller\Exception\MissingActionException If controller action is not found.
+     * @throws \UnexpectedValueException If return value of action method is not null or ResponseInterface instance.
+     */
+    public function invokeAction(): ?ResponseInterface;
```
If we normalize on ResponseInterface, linters will complain about the methods we've added. We should probably use Response.

ADmad replied:
Okay, let's keep the return type as they are for now.

markstory replied:
Done.

markstory committed:
Normalize ControllerInterface on ResponseInterface

This might require some additional casting hints in application code,
but these methods shouldn't frequently be used in applications.

markstory committed:
Revert type change.
